File: android/app/src/test/java/com/aurorastudios/authno/WidgetThemeTest.java
package com.aurorastudios.authno;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertTrue;

import org.junit.Test;

/**
 * WidgetTheme imports nothing from android, so it runs on the JVM.
 *
 * Worth pinning because it is the difference between a widget that matches the
 * app and one that renders somebody's Sepia theme as plain Dark — and because
 * everything it parses comes from JSON crossing a bridge, where a missing field
 * is not a hypothetical.
 */
public class WidgetThemeTest {

    // ── colour parsing ───────────────────────────────────────────────────────

    @Test public void sixDigitHex() {
        assertEquals(0xFF1A1B1E, WidgetTheme.color("#1a1b1e", 0));
        assertEquals(0xFFFFFFFF, WidgetTheme.color("#ffffff", 0));
    }

    @Test public void threeDigitHexExpands() {
        assertEquals(0xFFFFFFFF, WidgetTheme.color("#fff", 0));
        assertEquals(0xFF000000, WidgetTheme.color("#000", 0));
        assertEquals(0xFF112233, WidgetTheme.color("#123", 0));
    }

    @Test public void eightDigitHexKeepsItsAlpha() {
        assertEquals(0x80112233, WidgetTheme.color("#80112233", 0));
    }

    @Test public void whitespaceIsTolerated() {
        assertEquals(0xFF1A1B1E, WidgetTheme.color("  #1a1b1e  ", 0));
    }

    /**
     * The design system writes translucent surfaces as rgba(...). Those are
     * deliberately not parsed: a widget sits on an unknown wallpaper, so a
     * half-transparent card is worse than the fallback.
     */
    @Test public void rgbaFallsBackRatherThanRenderingTransparent() {
        assertEquals(0xFF123456, WidgetTheme.color("rgba(255,255,255,0.05)", 0xFF123456));
    }

    @Test public void junkFallsBack() {
        int fb = 0xFF123456;
        assertEquals(fb, WidgetTheme.color(null, fb));
        assertEquals(fb, WidgetTheme.color("", fb));
        assertEquals(fb, WidgetTheme.color("#", fb));
        assertEquals(fb, WidgetTheme.color("#12", fb));
        assertEquals(fb, WidgetTheme.color("#zzzzzz", fb));
        assertEquals(fb, WidgetTheme.color("1a1b1e", fb));   // no hash
    }

    // ── payload parsing ──────────────────────────────────────────────────────

    @Test public void aFullPayloadIsUsed() {
        WidgetTheme t = WidgetTheme.parse(
                "{\"bgColor\":\"#f4ecd8\",\"textPrimary\":\"#3b2f2f\","
                        + "\"textSecondary\":\"#5b4636\",\"textDim\":\"#8a7968\","
                        + "\"isDark\":false,\"themeId\":\"sepia\"}", true);
        assertEquals(0xFFF4ECD8, t.bg);
        assertEquals(0xFF3B2F2F, t.textPrimary);
        assertEquals(0xFF8A7968, t.textDim);
        assertFalse(t.isDark);
    }

    /** A widget placed before the app has ever synced must still render. */
    @Test public void noPayloadFallsBackToTheHint() {
        assertTrue(WidgetTheme.parse("", true).isDark);
        assertFalse(WidgetTheme.parse("", false).isDark);
        assertTrue(WidgetTheme.parse(null, true).isDark);
    }

    @Test public void malformedJsonFallsBackRatherThanThrowing() {
        WidgetTheme t = WidgetTheme.parse("{not json at all", true);
        assertTrue(t.isDark);
        assertEquals(WidgetTheme.fallback(true).bg, t.bg);
    }

    /**
     * A payload from a newer app version might omit a field this build reads.
     * Each one falls back on its own rather than losing the whole theme.
     */
    @Test public void aMissingFieldFallsBackOnItsOwn() {
        WidgetTheme t = WidgetTheme.parse("{\"bgColor\":\"#ffffff\",\"isDark\":false}", true);
        assertEquals(0xFFFFFFFF, t.bg);
        assertEquals(WidgetTheme.fallback(false).textPrimary, t.textPrimary);
        assertFalse(t.isDark);
    }

    /** The lightness in the payload wins over the older boolean. */
    @Test public void payloadLightnessOverridesTheHint() {
        assertFalse(WidgetTheme.parse("{\"isDark\":false}", true).isDark);
        assertTrue(WidgetTheme.parse("{\"isDark\":true}", false).isDark);
    }

    @Test public void fallbacksAreOpaque() {
        // A widget on an unknown wallpaper must not be see-through.
        assertEquals(0xFF000000, WidgetTheme.fallback(true).bg & 0xFF000000);
        assertEquals(0xFF000000, WidgetTheme.fallback(false).bg & 0xFF000000);
    }

    /**
     * StreakWidgetRenderer draws its calendar straight from these, so a role
     * left at 0 would paint transparent day numbers rather than fall back.
     */
    @Test public void everyCalendarRoleIsSet() {
        for (boolean dark : new boolean[]{true, false}) {
            WidgetTheme t = WidgetTheme.fallback(dark);
            for (int c : new int[]{t.bg, t.textPrimary, t.textSecondary, t.textDim,
                                   t.textFaint, t.textHasData, t.progressTrack,
                                   t.surface, t.surfaceRaised, t.border}) {
                assertEquals("opaque", 0xFF000000, c & 0xFF000000);
            }
        }
    }

    /** The config screen paints rows on these, so they must differ from the sheet. */
    @Test public void aRaisedSurfaceIsDistinctFromTheBackground() {
        for (boolean dark : new boolean[]{true, false}) {
            WidgetTheme t = WidgetTheme.fallback(dark);
            assertNotEquals(t.bg, t.surface);
            assertNotEquals(t.surface, t.surfaceRaised);
        }
    }

    @Test public void theCalendarRolesComeFromThePayload() {
        WidgetTheme t = WidgetTheme.parse(
                "{\"textFaint\":\"#111111\",\"textHasData\":\"#222222\","
                        + "\"progressTrack\":\"#333333\",\"isDark\":true}", true);
        assertEquals(0xFF111111, t.textFaint);
        assertEquals(0xFF222222, t.textHasData);
        assertEquals(0xFF333333, t.progressTrack);
    }

    /**
     * The design system's surface tokens are rgba(...) overlays. The progress
     * track has to be a solid colour for the same reason the card does, so the
     * app resolves it before sending; if it ever regresses to rgba the widget
     * must land on the fallback rather than draw nothing.
     */
    @Test public void anRgbaProgressTrackFallsBackToSolid() {
        WidgetTheme t = WidgetTheme.parse(
                "{\"progressTrack\":\"rgba(255,255,255,0.08)\",\"isDark\":true}", true);
        assertEquals(WidgetTheme.fallback(true).progressTrack, t.progressTrack);
    }

    // ── readableOn ───────────────────────────────────────────────────────────
    // The accent is the writer's own colour and every hue is allowed, so the
    // label on an accent-filled button cannot be hardcoded to white.

    @Test public void aDarkAccentTakesWhiteText() {
        assertEquals(0xFFFFFFFF, WidgetTheme.readableOn(0xFF5A00D9)); // the default violet
        assertEquals(0xFFFFFFFF, WidgetTheme.readableOn(0xFF000000));
        assertEquals(0xFFFFFFFF, WidgetTheme.readableOn(0xFF1A1B1E));
    }

    @Test public void aLightAccentTakesBlackText() {
        assertEquals(0xFF000000, WidgetTheme.readableOn(0xFFFFFFFF));
        assertEquals(0xFF000000, WidgetTheme.readableOn(0xFFFFC107)); // amber
        assertEquals(0xFF000000, WidgetTheme.readableOn(0xFF7FFF7F)); // pale green
    }

    /** Green dominates perceived luminance; blue barely registers. */
    @Test public void itWeighsChannelsPerceptually() {
        assertEquals(0xFF000000, WidgetTheme.readableOn(0xFF00FF00));
        assertEquals(0xFFFFFFFF, WidgetTheme.readableOn(0xFF0000FF));
    }

    /** Alpha in the accent must not be mistaken for a colour channel. */
    @Test public void alphaIsIgnored() {
        assertEquals(WidgetTheme.readableOn(0xFFFFFFFF), WidgetTheme.readableOn(0x00FFFFFF));
    }
}

File: src/theme/widgetTheme.test.js
/**
 * buildWidgetTheme is one half of a contract with Java.
 *
 * The other half is android/.../WidgetTheme.java, which parses #rgb, #rrggbb
 * and #aarrggbb and *deliberately refuses* rgba(...) — a widget sits on an
 * unknown wallpaper, so a translucent card would show the wallpaper through it.
 * Refusing means falling back, which means the app's theme silently does not
 * reach the home screen.
 *
 * That is exactly what happened to progressTrack, which was wired to
 * surfaces.mid — an rgba() token. Nothing in JS could notice: the value was a
 * perfectly good CSS colour. So the assertion has to live here, on the shape of
 * what crosses the bridge, for every theme that can be selected.
 */
import { buildWidgetTheme, flattenOver, onAccent } from './ThemeBase';
import { BUILTIN_THEMES } from './registry';
import { buildMaterialYouTheme } from './ThemeMaterialYou';

const THEMES = [...BUILTIN_THEMES, buildMaterialYouTheme()];

// Every field the Java side reads as a colour.
const COLOR_FIELDS = [
  'bgColor', 'textPrimary', 'textSecondary',
  'textDim', 'textFaint', 'textHasData', 'progressTrack',
  'surface', 'surfaceRaised', 'border',
];

/** The subset of CSS colours WidgetTheme.color() can actually parse. */
const PARSEABLE = /^#([0-9a-f]{3}|[0-9a-f]{6}|[0-9a-f]{8})$/i;

describe('buildWidgetTheme', () => {
  it('covers every selectable theme', () => {
    expect(THEMES.length).toBeGreaterThanOrEqual(6);
  });

  THEMES.forEach((theme) => {
    // The suite is generated from the theme registry on purpose — a theme
    // added later gets covered without anyone remembering to add a describe.
    // That makes the title dynamic, which is the only thing the rule objects to.
    // eslint-disable-next-line jest/valid-title
    describe(String(theme.meta?.id ?? 'unnamed'), () => {
      const w = buildWidgetTheme(theme);

      it('sends a colour the native parser accepts for every field', () => {
        COLOR_FIELDS.forEach((f) => {
          expect(typeof w[f]).toBe('string');
          expect(`${f}=${w[f]}`).toMatch(new RegExp(`^${f}=` + PARSEABLE.source.slice(1, -1) + '$', 'i'));
        });
      });

      it('never sends rgba(), which would fall back to the wrong theme', () => {
        COLOR_FIELDS.forEach((f) => {
          expect(String(w[f])).not.toMatch(/rgba?\(/i);
        });
      });

      it('states its own lightness rather than leaving it to a guess', () => {
        expect(typeof w.isDark).toBe('boolean');
        expect(w.isDark).toBe(theme.meta.isDark !== false);
      });

      it('carries an id so a payload can be traced back to a theme', () => {
        expect(w.themeId).toBe(theme.meta.id);
      });

      /**
       * The card and the text on it are painted from two different fields; if
       * they land on the same colour the widget renders blank.
       */
      it('does not paint text in the background colour', () => {
        ['textPrimary', 'textSecondary', 'textDim'].forEach((f) => {
          expect(String(w[f]).toLowerCase()).not.toBe(String(w.bgColor).toLowerCase());
        });
      });

      /**
       * The surfaces are the widget-config rows and the progress track. A
       * surface that composited to the background exactly would render the
       * whole list as one flat sheet.
       */
      it('keeps its raised surfaces distinguishable from the sheet', () => {
        ['surface', 'surfaceRaised', 'progressTrack', 'border'].forEach((f) => {
          expect(`${f}=${w[f]}`).not.toBe(`${f}=${w.bgColor}`);
        });
      });
    });
  });

  it('is JSON-round-trippable, because that is how it travels', () => {
    THEMES.forEach((theme) => {
      const w = buildWidgetTheme(theme);
      expect(JSON.parse(JSON.stringify(w))).toEqual(w);
    });
  });
});

describe('flattenOver', () => {
  it('composites an overlay the way the browser would', () => {
    // 50% white over black is mid grey.
    expect(flattenOver('rgba(255,255,255,0.5)', '#000000')).toBe('#808080');
    expect(flattenOver('rgba(0,0,0,0.5)', '#ffffff')).toBe('#808080');
  });

  it('leaves a fully opaque overlay alone', () => {
    expect(flattenOver('rgba(18,52,86,1)', '#ffffff')).toBe('#123456');
    expect(flattenOver('rgb(18,52,86)', '#ffffff')).toBe('#123456');
  });

  it('is a no-op at zero alpha', () => {
    expect(flattenOver('rgba(255,0,0,0)', '#123456')).toBe('#123456');
  });

  it('passes a solid straight through — nothing to composite', () => {
    expect(flattenOver('#abcdef', '#000000')).toBe('#abcdef');
  });

  /**
   * Returning the background makes the element invisible, which is the safe
   * failure: better than painting it in a colour nobody chose.
   */
  it('falls back to the background on anything it cannot read', () => {
    expect(flattenOver('hsl(200 50% 50%)', '#123456')).toBe('#123456');
    expect(flattenOver('rebeccapurple', '#123456')).toBe('#123456');
    expect(flattenOver(undefined, '#123456')).toBe('#123456');
    expect(flattenOver(null, '#123456')).toBe('#123456');
  });

  it('tolerates the whitespace real CSS carries', () => {
    expect(flattenOver('  rgba( 255 , 255 , 255 , 0.5 )  ', '#000000')).toBe('#808080');
  });
});

/**
 * onAccent decides the label colour on an accent-filled button, and the same
 * rule is implemented natively in WidgetTheme.readableOn — the two have to
 * agree or a button reads differently in the app and on the home screen.
 */
describe('onAccent', () => {
  const WHITE = '#ffffff';
  const INK   = '#111113';

  it('puts white on a dark accent', () => {
    expect(onAccent('#5a00d9')).toBe(WHITE); // the default violet
    expect(onAccent('#a855f7')).toBe(WHITE); // Violet
    expect(onAccent('#ec4899')).toBe(WHITE); // Rose
  });

  /**
   * The two shipped presets that were already unreadable in white — both sit
   * near 2:1 against it, well under the 4.5:1 that counts as readable. This is
   * the case that makes the whole helper worth having.
   */
  it('puts ink on the pale presets', () => {
    expect(onAccent('#f59e0b')).toBe(INK); // Gold
    expect(onAccent('#22c55e')).toBe(INK); // Sage
  });

  /**
   * These clear 3:1 in white and have read as white buttons since the app
   * shipped, so the rule leaves them alone. Pinned because the obvious
   * "maximise contrast" threshold would flip both, which is a redesign rather
   * than a fix.
   */
  it('leaves the mid-tone presets as they were', () => {
    expect(onAccent('#ff4500')).toBe(WHITE); // Ember, 3.4:1
    expect(onAccent('#3b82f6')).toBe(WHITE); // Ocean, 3.8:1
  });

  it('handles the extremes the colour wheel allows', () => {
    expect(onAccent('#ffffff')).toBe(INK);
    expect(onAccent('#000000')).toBe(WHITE);
    expect(onAccent('#00ff00')).toBe(INK);   // green dominates luminance
    expect(onAccent('#0000ff')).toBe(WHITE); // blue barely registers
  });

  it('does not throw on a missing or malformed accent', () => {
    expect(onAccent(undefined)).toBe(WHITE);
    expect(onAccent('')).toBe(WHITE);
    expect(onAccent('not a colour')).toBe(WHITE);
  });

  it('agrees with the native rule, which uses the same threshold', () => {
    // WidgetTheme.readableOn returns 0xFF000000 / 0xFFFFFFFF for these inputs;
    // pinned here so a change on one side fails on this side too.
    const nativeSaysBlack = ['#ffffff', '#f59e0b', '#22c55e', '#00ff00'];
    const nativeSaysWhite = ['#000000', '#5a00d9', '#1a1b1e', '#0000ff'];
    nativeSaysBlack.forEach((h) => expect(onAccent(h)).toBe(INK));
    nativeSaysWhite.forEach((h) => expect(onAccent(h)).toBe(WHITE));
  });
});
